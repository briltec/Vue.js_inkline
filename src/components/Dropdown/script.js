import { uid, isKey } from 'inkline/helpers';

import ClickOutside from 'inkline/directives/click-outside';

import AttributesProviderMixin from 'inkline/mixins/components/providers/AttributesProviderMixin';
import ClassesProviderMixin from 'inkline/mixins/components/providers/ClassesProviderMixin';
import EmitProviderMixin from 'inkline/mixins/components/providers/EmitProviderMixin';

import EmitFocusMethodMixin from 'inkline/mixins/components/methods/EmitFocusMethodMixin';

import VariantPropertyMixin from 'inkline/mixins/components/properties/VariantPropertyMixin';
import DisabledPropertyMixin from 'inkline/mixins/components/properties/DisabledPropertyMixin';

export default {
    name: 'IDropdown',
    mixins: [
        AttributesProviderMixin,
        ClassesProviderMixin,
        EmitProviderMixin,

        EmitFocusMethodMixin,

        VariantPropertyMixin,
        DisabledPropertyMixin
    ],
    directives: {
        ClickOutside
    },
    provide() {
        return {
            dropdown: this
        };
    },
    props: {
        trigger: {
            type: String,
            default: 'click'
        },
        hideOnClick: {
            type: Boolean,
            default: true
        },
        placement: {
            type: String,
            default: 'bottom'
        },
        showTimeout: {
            type: Number,
            default: 250
        },
        hideTimeout: {
            type: Number,
            default: 150
        }
    },
    data() {
        return {
            items: [],
            timeout: null,
            visible: false,
            triggerElement: null,
            dropdownElement: null,
            id: this.$attrs.id || uid('dropdown-menu')
        };
    },
    watch: {
        visible(value) {
            this.broadcast('IDropdownMenu', 'visibility-change', value);
            this.$emit('visibility-change', value);
        }
    },
    methods: {
        onTriggerKeyDown(e) {
            let activeIndex = this.items.findIndex((e) => e.active);
            let initialIndex = activeIndex > -1 ? activeIndex : 0;

            if (isKey('up', e) || isKey('down', e)) {
                this.show();

                setTimeout(() => {
                    this.items[initialIndex].$el.focus();
                }, this.visible ? 0 : this.showTimeout);

                e.preventDefault();
                e.stopPropagation();

            } else if (isKey('enter', e) || isKey('space', e)) {
                this.onClick();

                if (!this.visible) {
                    setTimeout(() => {
                        this.items[initialIndex].$el.focus();
                    }, this.showTimeout);
                }

                e.preventDefault();

            } else if (isKey('tab', e) || isKey('esc', e)) {
                this.hide();
            }
        },
        onItemKeyDown(e) {
            const target = e.target;
            const currentIndex = this.items.findIndex((i) => i.$el === e.target);
            const maxIndex = this.items.length - 1;
            let nextIndex;

            // Key: up || down
            if (isKey('up', e) || isKey('down', e)) {
                if (isKey('up', e)) {
                    nextIndex = currentIndex !== 0 ? currentIndex - 1 : 0;
                } else {
                    nextIndex = currentIndex < maxIndex ? currentIndex + 1 : maxIndex;
                }

                this.items[nextIndex].$el.focus();

                e.preventDefault();
                e.stopPropagation();

            } else if (isKey('enter', e) || isKey('space', e)) {
                target.click();

                this.triggerElement.focus();

                if (this.hideOnClick) {
                    this.visible = false;
                }

                e.preventDefault();

            } else if (isKey('tab', e) || isKey('esc', e)) {
                this.hide();

                this.triggerElement.focus();
            }
        },
        show() {
            if (this.disabled) return;

            clearTimeout(this.timeout);

            this.timeout = setTimeout(() => {
                this.visible = true;
            }, this.trigger === 'click' ? 0 : this.showTimeout);
        },
        hide() {
            if (this.disabled) return;

            clearTimeout(this.timeout);

            this.timeout = setTimeout(() => {
                this.visible = false;
            }, this.trigger === 'click' ? 0 : this.hideTimeout);
        },
        onClick() {
            if (this.disabled) return;

            if (this.visible) {
                this.hide();
            } else {
                this.show();
            }
        },
        initElements() {
            this.triggerElement = this.$slots.default[0].elm;
            this.dropdownElement = this.$slots.default[this.$slots.default.length - 1].elm;
        },
        initAriaAttributes() {
            this.dropdownElement.setAttribute('id', this.id);
            this.triggerElement.setAttribute('aria-haspopup', 'list');
            this.triggerElement.setAttribute('aria-controls', this.id);
        },
        initEvents() {
            this.triggerElement.addEventListener('keydown', this.onTriggerKeyDown);
            this.dropdownElement.addEventListener('keydown', this.onItemKeyDown, true);

            if (this.trigger === 'hover') {
                this.triggerElement.addEventListener('mouseenter', this.show);
                this.triggerElement.addEventListener('mouseleave', this.hide);
                this.dropdownElement.addEventListener('mouseenter', this.show);
                this.dropdownElement.addEventListener('mouseleave', this.hide);
            } else if (this.trigger === 'click') {
                this.triggerElement.addEventListener('click', this.onClick);
            }
        },
        handleMenuItemClick(action, instance) {
            if (this.hideOnClick) {
                this.visible = false;
            }

            this.$emit('action', action, instance);
        },
    },
    created() {
        this.$on('dropdown-item-mounted', (item) => {
            this.items.push(item);
        });

        this.$on('dropdown-item-destroyed', (item) => {
            this.items = this.items.filter((i) => i === item);
        });
    },
    mounted() {
        this.$on('menu-item-click', this.handleMenuItemClick);

        this.initElements();
        this.initEvents();
        this.initAriaAttributes();
    },
};
