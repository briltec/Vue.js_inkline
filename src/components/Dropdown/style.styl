.dropdown
    display: inline-block

.dropdown-menu > .dropdown
    width: 100%
    display: block