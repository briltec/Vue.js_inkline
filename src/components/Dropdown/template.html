<div class="dropdown" :class="classes['*']" v-click-outside={hide}>
    <slot name="trigger"/>
    <slot />
</div>