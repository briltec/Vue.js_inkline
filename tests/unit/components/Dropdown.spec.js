import { shallowMount } from '@vue/test-utils';
import Dropdown from 'inkline/components/Dropdown';

describe('Components', () => {
    describe('Dropdown', () => {
        let wrapper;

        beforeEach(() => {
            wrapper = shallowMount(Dropdown, {
                propsData: {
                    id: 'dropdown'
                },
                slots: {
                    default: ['<button/>', '<div/>']
                }
            });
        });

        it('should be named correctly', () => {
            expect(Dropdown.name).toEqual('IDropdown');
        });

        it('should render correctly', () => {
            expect(wrapper.html()).toMatchSnapshot();
        });

        describe('props', () => {
            describe('trigger', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.trigger).toBeDefined();
                    expect(wrapper.vm.trigger).toEqual('click');
                });
            });

            describe('hideOnClick', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.hideOnClick).toBeDefined();
                    expect(wrapper.vm.hideOnClick).toEqual(true);
                });
            });

            describe('placement', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.placement).toBeDefined();
                    expect(wrapper.vm.placement).toEqual('bottom');
                });
            });

            describe('showTimeout', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.showTimeout).toBeDefined();
                    expect(wrapper.vm.showTimeout).toEqual(250);
                });
            });

            describe('hideTimeout', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.hideTimeout).toBeDefined();
                    expect(wrapper.vm.hideTimeout).toEqual(150);
                });
            });
        });

        describe('data', () => {
            describe('items', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.items).toBeDefined();
                    expect(wrapper.vm.items).toEqual([]);
                });
            });

            describe('timeout', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.timeout).toBeDefined();
                    expect(wrapper.vm.timeout).toEqual(null);
                });
            });

            describe('visible', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.visible).toBeDefined();
                    expect(wrapper.vm.visible).toEqual(false);
                });
            });

            describe('triggerElement', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.triggerElement).toBeDefined();
                });
            });

            describe('dropdownElement', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.dropdownElement).toBeDefined();
                });
            });

            describe('id', () => {
                it('should be defined', () => {
                    expect(wrapper.vm.id).toBeDefined();
                    expect(wrapper.vm.id).toEqual('dropdown');
                });

                it('should be randomly generated if not defined', () => {
                    wrapper = shallowMount(Dropdown, {
                        slots: {
                            default: ['<button/>', '<div/>']
                        }
                    });

                    expect(wrapper.vm.id).toMatch(/^dropdown-\w+/);
                });
            });
        });

        describe('watch', () => {
            describe('visible', () => {
                it('should broadcast "visibility-change" to IDropdownMenu', () => {
                    const spy = jest.spyOn(wrapper.vm, 'broadcast');

                    wrapper.setData({
                        visible: true
                    });

                    expect(spy).toHaveBeenCalled();
                    expect(spy).toHaveBeenCalledWith('IDropdownMenu', 'visibility-change', true);
                });

                it('should emit "visibility-change"', () => {
                    const spy = jest.spyOn(wrapper.vm, '$emit');

                    wrapper.setData({
                        visible: true
                    });

                    expect(spy).toHaveBeenCalled();
                    expect(spy).toHaveBeenCalledWith('visibility-change', true);
                });
            });
        });

        describe('methods', () => {
            describe('onTriggerKeyDown()', () => {
                [
                    'ArrowUp',
                    'ArrowDown'
                ].forEach((key) => {
                    const e = {
                        key,
                        preventDefault: () => {},
                        stopPropagation: () => {}
                    };

                    it.only('should call show() if "' + key + '" key pressed', () => {
                        const spy = jest.spyOn(wrapper.vm, 'show');

                        wrapper.vm.onTriggerKeyDown(e);

                        expect(spy).toHaveBeenCalled();
                    });

                    it('should call focus() on item at 0 index if "' + key + '" key pressed', (done) => {
                        wrapper = shallowMount(Dropdown, {
                            propsData: {
                                id: 'dropdown'
                            },
                            data() {
                                return {
                                    items: [
                                        {
                                            $el: { focus: () => {} }
                                        }
                                    ]
                                };
                            },
                            slots: {
                                default: ['<button/>', '<div/>']
                            }
                        });

                        const spy = jest.spyOn(wrapper.vm.items[0].$el, 'focus');

                        wrapper.vm.onTriggerKeyDown(e);

                        setTimeout(() => {
                            expect(spy).toHaveBeenCalled();
                            done();
                        }, 500);
                    });

                    it('should call focus() on item at initialIndex if "' + key + '" key pressed', (done) => {
                        wrapper = shallowMount(Dropdown, {
                            propsData: {
                                id: 'dropdown'
                            },
                            data() {
                                return {
                                    items: [
                                        {
                                            $el: { focus: () => {} }
                                        },
                                        {
                                            active: true,
                                            $el: { focus: () => {} }
                                        }
                                    ]
                                };
                            },
                            slots: {
                                default: ['<button/>', '<div/>']
                            }
                        });

                        const spy = jest.spyOn(wrapper.vm.items[1].$el, 'focus');

                        wrapper.vm.onTriggerKeyDown(e);

                        setTimeout(() => {
                            expect(spy).toHaveBeenCalled();
                            done();
                        }, 500);
                    });

                    it('should call event preventDefault() and stopPropagation() if "' + key + '" key pressed', () => {
                        const spy1 = jest.spyOn(e, 'preventDefault');
                        const spy2 = jest.spyOn(e, 'stopPropagation');

                        wrapper.vm.onTriggerKeyDown(e);

                        expect(spy1).toHaveBeenCalled();
                        expect(spy2).toHaveBeenCalled();
                    });
                });
            });
        });
    });
});
