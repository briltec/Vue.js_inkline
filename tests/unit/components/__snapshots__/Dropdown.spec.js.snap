// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Components Dropdown should render correctly 1`] = `
<div id="dropdown" class="dropdown"><button aria-haspopup="list" aria-controls="dropdown"></button>
  <div id="dropdown"></div>
</div>
`;
